File: web/src/type/general.ts
import {CSSProperties, ReactElement} from "react";
import {Theme} from "@mui/material";
import {SecretStatus} from "./secret";
import {SystemStyleObject} from "@mui/system/styleFunctionSx/styleFunctionSx";

// COMMON (WEB AND SERVER)

export interface Response<TData, TError = object> {
    response: TData,
    error: TError,
}

export interface Login {
    username: string,
    password: string,
}

export interface DbConnection {
    host: string,
    port: number,
    credentialId: string,
}

export enum AuthType {
    NONE,
    BASIC,
}

export interface AppInfo {
    company: string,
    configured: boolean,
    secret: SecretStatus,
    availability: Availability,
    version: Version,
    auth: AuthInfo,
}

export interface AuthInfo {
    authorised: boolean,
    error: string,
    type: AuthType,
}

export interface AppConfig {
    company: string,
    availability: Availability,
    auth: AuthConfig,
}

export interface AuthConfig {
    type: AuthType,
    body: any,
}

export interface Availability {
    manualQuery: boolean,
}

export interface Version {
    tag: string,
    commit: string,
    label: string,
}

export enum FileUsageType {
    UPLOAD,
    PATH,
}

export interface Database {
    host: string,
    port: number,
    name?: string,
}

export enum SidecarStatus {
    Active = "ACTIVE",
    Paused = "PAUSED",
}
export interface Sidecar {
    host: string,
    port: number,
    status?: SidecarStatus,
}

// SPECIFIC (WEB)

export interface StylePropsMap {
    [key: string]: CSSProperties,
}

export interface SxPropsMap {
    // NOTE: we don't need usage of array here that is why it is not use SxProps<Theme> interface
    [key: string]: SystemStyleObject<Theme> | ((theme: Theme) => SystemStyleObject<Theme>),
}

export interface ColorsMap {
    [name: string]: "success" | "primary" | "error" | "warning",
}

export interface EnumOptions {
    label: string,
    key: string,
    name?: string,
    icon: ReactElement,
    color?: string,
    badge?: string,
}

export enum Settings {
    MENU,
    PASSWORD,
    CERTIFICATE,
    SECRET,
    ABOUT,
}

export interface Links {
    [key: string]: {
        name: string,
        link: string,
    }
}

File: service/src/model/general.go
package model

import (
	"github.com/google/uuid"
)

// COMMON (WEB AND SERVER)

type Response struct {
	Response any `json:"response"`
	Error    any `json:"error"`
}

type Login struct {
	Username string `form:"username" json:"username,omitempty"`
	Password string `form:"password" json:"password,omitempty"`
}

type DbConnection struct {
	Host         string    `json:"host"`
	Port         int       `json:"port"`
	CredentialId uuid.UUID `json:"credentialId"`
}

type AuthType int8

const (
	NONE AuthType = iota
	BASIC
)

type AppInfo struct {
	Company      string       `json:"company"`
	Configured   bool         `json:"configured"`
	Availability Availability `json:"availability"`
	Secret       SecretStatus `json:"secret"`
	Version      Version      `json:"version"`
	Auth         AuthInfo     `json:"auth"`
}

type AuthInfo struct {
	Type       AuthType `json:"type"`
	Authorised bool     `json:"authorised"`
	Error      string   `json:"error"`
}

type AppConfig struct {
	Company      string       `json:"company"`
	Availability Availability `json:"availability"`
	Auth         AuthConfig   `json:"auth"`
}

type AuthConfig struct {
	Type AuthType          `json:"type"`
	Body map[string]string `json:"body"`
}

type Availability struct {
	ManualQuery bool `json:"manualQuery"`
}

type Version struct {
	Tag    string `json:"tag"`
	Commit string `json:"commit"`
	Label  string `json:"label"`
}

type FileUsageType int8

const (
	UPLOAD FileUsageType = iota
	PATH
)

type Database struct {
	Host string  `json:"host"`
	Port int     `json:"port"`
	Name *string `json:"name"`
}

type SidecarStatus string

const (
	Paused SidecarStatus = "PAUSED"
	Active               = "ACTIVE"
)

type Sidecar struct {
	Host   string         `json:"host" form:"host"`
	Port   int            `json:"port" form:"port"`
	Name   *string        `json:"name" form:"name"`
	Status *SidecarStatus `json:"status,omitempty" form:"status"`
}

// SPECIFIC (SERVER)
